text/3-entrees/storage-api/vec-set.md

```diff
@@ -0,0 +1,145 @@
+		// they should be inserted
+		Err(index) => {
+			members.insert(index, new_member.clone());
+			Members::<T>::put(members);
+			Self::deposit_event(RawEvent::MemberAdded(new_member));
+			Ok(())
+		}
+	}
+}
+```
+
+If it turns out that the caller is not already a member, the binary search will fail. In this case
+it still gives back the index into the `Vec` at which the member would have been stored had he been
+present. We then use this information to insert the member at the appropriate location, thus
+maintaining a sorted `Vec`.
+
+## Removing a Member
+
+Removing a member is entirely straightforward. We begin by looking for the caller in the list. If
+the caller is not present, there is no work to be done. If the caller is present, the search
+algorithm returns her index, and she can be removed.
+
+```rust, ignore
+fn remove_member(origin) -> DispatchResult {
+	let old_member = ensure_signed(origin)?;
+
+	let mut members = Members::<T>::get();
+
+	// We have to find out where, in the sorted vec the member is, if anywhere.
+	match members.binary_search(&old_member) {
+		// If the search succeeds, the caller is a member, so remove her
+		Ok(index) => {
+			members.remove(index);
+			Members::<T>::put(members);
+			Self::deposit_event(RawEvent::MemberRemoved(old_member));
+			Ok(())
+		},
+		// If the search fails, the caller is not a member, so just return
+		Err(_) => Err(Error::<T>::NotMember.into()),
+	}
+}
+```
+
+## Performance
+
+Now that we have built our set, let's analyze its performance in some common operations.
+
+### Membership Check
+
+In order to check for the presence of an item in a vec set, we make a single storage read, decode
+the entire vector, and perform a binary search.
+
+DB Reads: O(1) Decoding: O(n) Search: O(log n)
+
+### Updating
+
+Updates to the set, such as adding and removing members, like we demonstrated, requires first
+performing a membership check. In addition it requires re-encoding the entire `Vec` and storing back
+in the database.
+
+DB Writes: O(1) Encoding: O(n)
+
+### Iteration
+
+Iterating over all items in a `vec-set` is achieved by using the `Vec`'s own
+[`iter` method](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.iter). The entire set can,
+again, be read from storage in one go, and each item must be decoded. Finally, the actual processing
+you do on the items will take some time.
+
+DB Reads: O(1) Decoding: O(n) Processing: O(n)
+
+Because accessing the database is a relatively slow operation, only executing a single read while
+iterating the entire list is a big win. If your data will be iterated frequently, you may want a
+`vec-set`.
+
+### A Note on Weights
+
+It is always important that the weight associated with your dispatchables represent the actual time
+it takes to execute them. In this pallet we have provided an upper bound on the size of the set,
+which places an upper bound on the computation. Thus we get away with constant weight annotations.
```

Comment:
```suggestion
it takes to execute them. In this pallet, we have provided an upper-bound on the size of the set,
which places an upper-bound on the computation - this means we can use constant weight annotations.
```

Reply:
I think `upper bound` is correct, otherwise 👍

Reply:
You are correct...thank you!